### ISSUE TITLE
concatSubstreams hangs when used as documented.

### ISSUE THREAD

**Issue by BarDweller** [akka/akka#25814]
I'm trying to use concatSubstreams to collapse the results of a groupBy in order (as I can do with rxjava/reactor etc). But when I try this with akka, it hangs after outputting the first element. The docs do indicate it could hang if the number of groupBy streams were insufficient, but the docs also include an example, which I expected to work. 

The docs reference this code in the test suite.. 
https://github.com/akka/akka/blob/v2.5.17/akka-docs/src/test/java/jdocs/stream/SubstreamDocTest.java#L67-L70

Here's that example from the docs/test case slightly modified to run standalone (added println sink) 
```java
final ActorSystem system = ActorSystem.create("QuickStart");
final Materializer mat = ActorMaterializer.create(system);
Source.from(Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9, 10))
  .groupBy(3, elem -> elem % 3)
  .concatSubstreams()
  .runWith(Sink.foreach(a -> System.out.println(a)), mat);
```

When this runs I would expect to see 
```
1
4
7
10
2
5
8
3
6
9
```

But instead I just see `1` then it hangs. 

Are the docs incorrect? are groupBy/concatSubstreams defective? or am I missing something obvious as to why it isn't running?

**Comment by johanandren** [comment]
Thanks for letting us know.

That sample is not correct, the scaladoc/javadocs on the `concatSubstream` method says:

> This is usually a bad idea when combined with `groupBy` since it can
> easily lead to deadlock—the concatenation does not consume from the second
> substream until the first has finished and the `groupBy` operator will get
> back-pressure from the second stream.

Which is exactly what is happening with that sample.

Actually not quite sure what would be a good sample for it, but we should definitely not have something that wrong like that in the docs (it is wrong for both the Java and Scala sample)

**Comment by hepin1989** [comment]
@BarDweller I was surprised too, I was expected it would work as a charm after I added the `grouped(10` after the `groupBy` call.

**Comment by hepin1989** [comment]
@BarDweller I was expected it would work if you add a `grouped(10)` call after that, but it doesn't.

**Comment by BarDweller** [comment]
@johanandren I would also be concerned about why the test case that has the same code as that sample is not causing build failures. At a guess because it's the last test in that block and it doesn't throw an exception, it's been considered as passing even tho it has hung. It looks like those tests are primarily there to check the example code is valid & compiles, but it's noteworthy that none of them attempt to verify that the behavior is as expected, just sending to Sink.ignore

**Comment by johanandren** [comment]
Yeah, it is more of a compilation test as it looks now. It can sometimes be hard to write code that is  verifiable without making it completely useless as a standalone sample. I think this should be possible to run with a `Sink.seq` and verify the collected elements though, so that is another improvement to do right away if anyone picks this up.

**Comment by BarDweller** [comment]
I still don't know how I'm supposed to achieve this in akka, if it should have worked with a `grouped(10)` added, is this actually a bug (as well as a doc issue) ?

**Comment by johanandren** [comment]
This is not a bug, it is how those two operators work in concert. `concatSubstreams` means to concatenate one sub stream at a time downstream, until it completes, and backpressure all other sub streams, while `groupBy` cannot pull further elements from upstream while there is backpressure as it does not know beforehand which downstream it will end up in.

By adding some form of detaching/aggregating operator downstream of `groupBy` you make sure there is always demand even if `concatSubstreams` is back pressuring.

The bug is that the documentation show this incorrect example.

**Comment by BarDweller** [comment]
I understand why (with no intermediate operators) this is going to deadlock due to backpressure..

But the suggestion has been made that adding `grouped(10)` should have let it work, when it doesn't.

**Comment by johanandren** [comment]
Perhaps not obvious, but `grouped` wont start pulling into its buffer until there is downstream demand, which means that the first substream with grouped will have demand, but for value 2 there will already be backpressure so the `groupBy` will start backpressuring. 

Looks like the docs for `grouped` is not clear about this though, I created #25909 to track that.

**Comment by hepin1989** [comment]
@BarDweller `groupBy` followed by `mergeSubstreamsWithParallelism(1)` here do not work...I don't know how to get the expected result, even with `buffer`. change to `1` to `3` do consume, but can't get what you wanted.

**Comment by BarDweller** [comment]
@johanandren Ok, so you're saying that the docs for `concatSubstreams` AND `grouped` both aren't clear enough to let me figure out how to do the akka equivalent of something that's trivial in both rxJava and WebFlux. 

Can _you_ help provide an example of how akka is supposed to be used to stream a set of entities (let's assume it's a fixed set, eg from `Source.range` as opposed to a potentially infinite stream).. and feed that stream through `groupBy` to perform processing on substreams identified by an appropriate function.. and then subsequently flattened to a single stream with the groups being aggregated together ? 

Because all I'm seeing at the mo as part of this issue is "It's not a bug, the docs are wrong" along with "but I can't make it work either". 

I can supply the rx / webflux code if it would be useful.

**Comment by johanandren** [comment]
I'm not sure I understand well enough what you want to achieve, but please do open up a discussion in the forums ( https://discuss.akka.io ) rather than continuing that discussion in the bug tracker, thanks!
